contracts/consensus/Coconsensus.sol

```diff
@@ -0,0 +1,366 @@
+     */
+    function setup() public {
+
+        require(
+            selfMetachainId == bytes32(0),
+            "Coconsensus contract is already initialized."
+        );
+
+        originMetachainId = genesisOriginMetachainId;
+
+        selfMetachainId = genesisSelfMetachainId;
+
+        relativeSelfDynasty = uint256(0);
+
+        bytes32 currentMetachainId = genesisMetachainIds[SENTINEL_METACHAIN_ID];
+
+        // Loop through the genesis metachainId link list.
+        while (currentMetachainId != SENTINEL_METACHAIN_ID) {
+
+            // Setup protocore contract for the given metachain id.
+            setupProtocore(currentMetachainId);
+
+            // Setup observer contract for the given metachain id.
+            setupObserver(currentMetachainId);
+
+            // Traverse to next metachain id from the link list mapping.
+            currentMetachainId = genesisMetachainIds[currentMetachainId];
+        }
+    }
+
+
+    /* External Functions */
+
+    /**
+     * @notice finaliseCheckpoint() function finalizes a checkpoint at
+     *         a metachain.
+     *
+     * @param _metachainId A metachain id to finalize a checkpoint.
+     * @param _blockNumber A block number of a checkpoint.
+     * @param _blockHash A block hash of a checkpoint.
+     *
+     * \pre `_metachainId` is not 0.
+     * \pre `_blockHash` is not 0.
+     * \pre `msg.sender` should be the protocore contract.
+     * \pre `_blockNumber` must be multiple of epoch length.
+     * \pre `_blockNumber` must be greater than the last finalized block number.
+     *
+     * \post Increment the `relativeSelfDynasty` storage value by one if the
+     *       `msg.sender` is self protocore contract address.
+     * \post Adds a new `Block` in the `blockchains` mapping.
+     * \post Updates the `blockTips` mapping with provided `_blockNumber`.
+     */
+    function finaliseCheckpoint(
+        bytes32 _metachainId,
+        uint256 _blockNumber,
+        bytes32 _blockHash
+    )
+        external
+        onlyRunningProtocore(_metachainId)
+    {
+        // Check if the metachain id is not null.
+        require(
+            _metachainId != bytes32(0),
+            "Metachain id must not be null."
+        );
+
+        // Check if the blockhash is not null.
+        require(
+            _blockHash != bytes32(0),
+            "Blockhash must not be null."
+        );
+
+        /*
+         * Check if the new block number is greater than the last
+         * finalised block number.
+         */
+        uint256 lastFinalisedBlockNumber = blockTips[_metachainId];
+        require(
+            _blockNumber > lastFinalisedBlockNumber,
+            "The block number of the checkpoint must be greater than the block number of last finalised checkpoint."
+        );
+
+        // Check if the block number is multiple of epoch length.
+        ProtocoreI protocore = protocores[_metachainId];
+        uint256 epochLength = protocore.epochLength();
+        require(
```

Comment:
I believe, this should be either removed (we trust implementation in Protocore) or changed to an assert.

Reply:
Changed to assert.